src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,353 @@
+                        c => UpdateDocumentAsync(document, span, ConvertToRawKind.MultiLineIndented, c),
+                        nameof(CSharpFeaturesResources.Convert_to_raw_string),
+                        priority),
+                    token.Span);
+
+                if (lineOffset > 0)
+                {
+                    context.RegisterRefactoring(
+                        new MyCodeAction(
+                            CSharpFeaturesResources.Convert_to_raw_string_no_indent,
+                            c => UpdateDocumentAsync(document, span, ConvertToRawKind.MultiLine, c),
+                            nameof(CSharpFeaturesResources.Convert_to_raw_string_no_indent),
+                            priority),
+                        token.Span);
+                }
+            }
+        }
+
+        private static bool CanBeSingleLine(VirtualCharSequence characters)
+        {
+            // Single line raw strings cannot start/end with quote.
+            if (characters.First().Rune.Value == '"' ||
+                characters.Last().Rune.Value == '"')
+            {
+                return false;
+            }
+
+            // a single line raw string cannot contain a newline.
+            if (characters.Any(static ch => IsNewLine(ch)))
+                return false;
+
+            return true;
+        }
+
+        private static bool IsNewLine(VirtualChar ch)
+            => ch.Rune.Utf16SequenceLength == 1 && SyntaxFacts.IsNewLine((char)ch.Value);
+
+        private static bool AllEscapesAreQuotes(VirtualCharSequence sequence)
+        {
+            var hasEscape = false;
+
+            foreach (var ch in sequence)
+            {
+                if (ch.Span.Length > 1)
+                {
+                    hasEscape = true;
+                    if (ch.Value != '"')
+                        return false;
+                }
+            }
+
+            return hasEscape;
+        }
+
+        private static bool IsInDirective(SyntaxNode? node)
+        {
+            while (node != null)
+            {
+                if (node is DirectiveTriviaSyntax)
+                    return true;
+
+                node = node.GetParent(ascendOutOfTrivia: true);
+            }
+
+            return false;
+        }
+
+        private static bool CanConvert(VirtualChar ch)
+        {
+            // Don't bother with unpaired surrogates.  This is just a legacy language corner case that we don't care to
+            // even try having support for.
+            if (ch.SurrogateChar != 0)
+                return false;
+
+            // Can't ever encode a null value directly in a c# file as our lexer/parser/text apis will stop righ there.
+            if (ch.Rune.Value == 0)
+                return false;
+
+            // Check if we have an escaped character in the original string.
+            if (ch.Span.Length > 1)
+            {
+                // An escaped newline is fine to convert (to a multi-line raw string).
```

Comment:
This seems like a very Windows-centric assumption. What characters does `IsNewLine` match? Because, and maybe I missed them, but I didn't see tests for other newline characters like `\n` and `\r` on their own.

I think this needs a `(may change semantics)` (or some other wording) note, because this converts specifically defined newline character sequences into whatever the current machine/git repo/etc. happens to be configured to use. If a user checks out a git repo on a mac, runs this refactoring on VS Mac, this could break things.

Unless I'm wrong and the new raw strings exclusively encode `\r\n` into the value? If so, then I guess this is okay, but would be good to confirm that `IsNewLine` doesn't match anything other than that.

Reply:
So this is a good point.  For clarity, whatever newline sequence you wrote, we will embed into your converted string.  This means from the language/compiler perspective, on that machine, the literal has the exact meaning as before (no semantics change).

Where it gets subtle though is the absolutely insane behavior of git where it may take a file and just change newlines *even that are language relevant* when sent to *another* machine.

In this case, the refactoring is in line with the lang and compiler.  However, it is *git* that ends up causing issues here as it thinks it is sensible to touch programming lines that relate to content.

--

Note that this is a refactoring though.  It is opted into the user, presumably because they actually think this is better.  And it will be very clear that a single line became multi-line.  So i think the user can be trusted to decide if that is acceptable for them given their git config.

Reply:
> For clarity, whatever newline sequence you wrote, we will embed into your converted string. This means from the language/compiler perspective, on that machine, the literal has the exact meaning as before (no semantics change).

Read the code now, and yes, I can see that. I think thats probably good, and perhaps even the saviour of it. If I was on a system that used LF, and converted a string that had CRLF, I would expect the IDE would somehow alert me to the fact that I had mixed line endings, and I could easily Ctrl+Z.